include/networkit/graph/Graph.hpp

```diff
@@ -139,6 +145,389 @@ class Graph final {
     //!< same schema (and same order!) as outEdges
     std::vector<std::vector<edgeid>> outEdgeIds;
 
+private:
+    // user attributes of this graph (node attributes so far)
+    // Base class for all node attributes.
+    class NodeAttributeStorageBase {
+    public:
+        NodeAttributeStorageBase(const Graph *graph, std::string name, std::type_index type)
+            : name{std::move(name)}, type{type}, theGraph{graph} {}
+
+        virtual ~NodeAttributeStorageBase() = default;
+
+        virtual void invalidateAttributes() = 0;
+
+        std::string getName() { return name; }
+
+        std::type_index getType() { return type; }
+
+        bool isValid(index i) { return i < valid.size() && valid[i]; }
+
+        // Called by Graph when node is deleted.
+        void invalidate(index i) {
+            if (isValid(i)) {
+                valid[i] = false;
+                --validElements;
+            }
+        }
+
+    protected:
+        void markValid(index i) {
+            if (!theGraph->hasNode(i)) {
+                throw std::runtime_error("This node does not exist");
+            }
+            if (i >= valid.size()) {
+                valid.resize(i + 1);
+            }
+            valid[i] = true;
+            ++validElements;
+        }
+
+        void checkIndex(index i) {
+            if (!theGraph->hasNode(i)) {
+                throw std::runtime_error("This node does not exist");
+            }
+            if (!isValid(i)) {
+                throw std::runtime_error("Invalid attribute value");
+            }
+        }
+
+    private:
+        std::string name;
+        std::type_index type;
+        std::vector<bool> valid; // For each node: whether attribute is set or not.
+    protected:
+        index validElements = 0;
+        const Graph *theGraph;
+    }; // class NodeAttributeStorageBase
+
+    template <typename T>
+    class NodeAttribute;
+
+    template <typename T>
+    class NodeAttributeStorage : public Graph::NodeAttributeStorageBase {
+    public:
+        NodeAttributeStorage(const Graph *theGraph, std::string name)
+            : NodeAttributeStorageBase{theGraph, std::move(name), typeid(T)} {}
+
+        ~NodeAttributeStorage() override { invalidateAttributes(); }
+
+        void invalidateAttributes() override {
+            for (auto att : attrSet)
+                att->invalidateAttribute();
+        }
+
+        void resize(index i) {
+            if (i >= values.size()) {
+                values.resize(i + 1);
+            }
+        }
+
+        auto size() { return validElements; }
+
+        void set(index i, T v) {
+            markValid(i);
+            resize(i);
+            values[i] = std::move(v);
```

Comment:
For this to work, shouldn't `v` be an rvalue ref? `void set(index i, T &&v)`.

Reply:
Still still works (and moves from the local `v` to `values`). It just doesn't move from the caller's argument.